EngineLayer/FdrAnalysis/PEPValueAnalysisGeneric.cs

```diff
@@ -57,84 +57,150 @@ public static string ComputePEPValuesForAllPSMsGeneric(List<PeptideSpectralMatch
+            double logLossReductionAverage = Double.NegativeInfinity;
+
+            if ((logLoss != null) && (logLoss.Any()))
             {
-                maxThreads = 1;
+                logLossAverage = logLoss.Average();
             }
 
-            int ambiguousPeptidesRemovedCount = 0;
+            if ((logLossReduction != null) && (logLossReduction.Any()))
+            {
+                logLossReductionAverage = logLossReduction.Average();
+            }
+
+            StringBuilder s = new StringBuilder();
+            s.AppendLine();
+            s.AppendLine("************************************************************");
+            s.AppendLine("*       Metrics for Determination of PEP Using Binary Classification      ");
+            s.AppendLine("*-----------------------------------------------------------");
+            s.AppendLine("*       Accuracy:  " + accuracy.Average().ToString());
+            s.AppendLine("*       Area Under Curve:  " + areaUnderRocCurve.Average().ToString());
+            s.AppendLine("*       Area under Precision recall Curve:  " + areaUnderPrecisionRecallCurve.Average().ToString());
+            s.AppendLine("*       F1Score:  " + F1Score.Average().ToString());
+            s.AppendLine("*       LogLoss:  " + logLossAverage.ToString());
+            s.AppendLine("*       LogLossReduction:  " + logLossReductionAverage.ToString());
+            s.AppendLine("*       PositivePrecision:  " + positivePrecision.Average().ToString());
+            s.AppendLine("*       PositiveRecall:  " + positiveRecall.Average().ToString());
+            s.AppendLine("*       NegativePrecision:  " + negativePrecision.Average().ToString());
+            s.AppendLine("*       NegativeRecall:  " + negativeRecall.Average().ToString());
+            s.AppendLine("*       Count of Ambiguous Peptides Removed:  " + sumOfAllAmbiguousPeptidesResolved.ToString());
+            s.AppendLine("************************************************************");
+            return s.ToString();
 
-            Parallel.ForEach(Partitioner.Create(0, psms.Count),
+        }
+
+        public static int Compute_PSM_PEP(List<PeptideSpectralMatch> psms, List<int> psmIndices, MLContext mLContext, TransformerChain<BinaryPredictionTransformer<Microsoft.ML.Calibrators.CalibratedModelParametersBase<Microsoft.ML.Trainers.FastTree.FastTreeBinaryModelParameters, Microsoft.ML.Calibrators.PlattCalibrator>>> trainedModel, string searchType, List<(string fileName, CommonParameters fileSpecificParameters)> fileSpecificParameters, Dictionary<string, int> sequenceToPsmCount, Dictionary<string, float> fileSpecificMedianFragmentMassErrors, int chargeStateMode, string outputFolder)
+        {
+            int maxThreads = fileSpecificParameters.FirstOrDefault().fileSpecificParameters.MaxThreadsToUsePerFile;
+            object lockObject = new object();
+            int ambiguousPeptidesResolved = 0;
+
+            if (String.IsNullOrEmpty(outputFolder))
```

Comment:
This (and ln 187) feels hacky. Can you comment what's going on here?

Reply:
i think it is. but i added a comment to explain. definitely up for alternatives.